docs/RFCS/20171118_role_based_access_control.md

```diff
@@ -0,0 +1,747 @@
+* show all memberships for a user: `SHOW GRANT ON ROLE * FOR marc`
+* show user membership in a role: `SHOW GRANT ON ROLE myrole FOR marc`
+
+### Administrators
+
+Before this RFC, there is a single administrative user in a cockroach cluster: the user `root`.
+
+We propose to add a default role: `admin` with a single user: `root`.
+`root` will have admin privileges on the `admin` role, allowing addition/removal of other users.
+
+The `admin` role cannot be deleted, and cannot be empty. The special `root` user cannot be dropped or removed
+from the `admin` role.
+
+All operations that previously checked for the `root` user will now check for membership in the `admin` role.
+
+Clusters created before the existence of roles will have a new `admin` role created automatically and privileges
+for it added where `root` currently has privileges. `root` will keep its individual privileges to allow downgrades
+to older versions.
+
+### Non-enterprise users
+
+Manipulation of roles is an enterprise feature.
+
+Without a valid enterprise license, the following are **not** allowed:
+* creating a role (`CREATE ROLE ...`)
+* dropping a role (`DROP ROLE ...`)
+* adding a member of a role (`GRANT myrole TO ...`)
+* removing a member from a role (`REVOKE myrole TO ...`)
+
+The following will function without an enterprise license:
+* listing roles
+* listing role memberships
+* granting privileges to a role
+* revoking privileges from a role
+
+This restriction leaves non-enterprise users with a fully-functioning product but without the
+ease of management provided by roles.
+
+## Contributor impact
+
+Any modifications to the SQL code must keep in mind:
+* privileges can be applied to roles and users
+* roles and users share the same namespace
+
+# Reference-level explanation
+
+A number of concepts are detailed in the [Guide-level explanation](#guide-level-explanation) and not
+repeated in this section.
+
+## Detailed design
+
+### Roles system table
+
+We propose the modification of the users table to store roles, and the addition of a new table to store
+role memberships.
+
+#### Users table
+
+Since roles and users share the same namespace, it is simplest to alter the `system.users` table to
+include roles.
+
+The schema for the users table becomes:
+```
+CREATE TABLE system.users (
+  username         STRING PRIMARY KEY,
+  "hashedPassword" BYTES,
+  isRole           BOOL,
+  (INDEX isRole)
+);`
+```
+
+Depending on the complexity of migration, we may be able to rename the `username` field and even the table
+to `rolename` and `system.roles` to reflect the fact that users are really just roles.
+
+`isRole` denotes whether the corresponding name is a user or a role. Roles cannot log in or have passwords.
+
+#### Role members table
+
+The `role_members` stores the list of all role memberships. This only stores direct relationships:
+
+```
+CREATE TABLE system.role_members (
+  role          STRING,
+  member        STRING,
```

Comment:
I take it that renaming a user/role isn't possible? If it was I'd expect to see IDs be used for membership.

Reply:
correct. That would be the `alter` command which is not currently supported. Adding details along with other things about `alter` in out-of-scope section.
We could use IDs, but two things make this undesirable:
* migrations are already tricky, re-keying the users table would be horrible
* rename is rare and is not in the path of regular queries so it's ok to be slow

Reply:
Well we could get around the re-keying issue by just adding an ID column with a secondary index on it. You're right that rename is rare, we'll just need to have very good testing to make sure we don't leave any references around to the old name. Things like foreign table references and view dependencies have had bugs in their initial implementations.